File: frontend/src/types/grading.ts
export interface RubricCriterion {
  id: string;
  name: string;
  description: string;
  maxScore: number;
  weight: number;
  orderIndex: number;
}

export interface Rubric {
  id: string;
  name: string;
  description: string | null;
  rubricType: 'built_in' | 'custom';
  createdAt: string;
  criteria: RubricCriterion[];
}

export interface CriterionScore {
  criterion_id: string;
  criterion_name: string;
  score: number;
  max_score: number;
  feedback: string;
}

export interface FillerWord {
  word: string;
  count: number;
}

export interface PacingSegment {
  start: number;
  end: number;
  wpm: number;
}

export interface GradingDetailedResults {
  criterion_scores: CriterionScore[];
  filler_words: FillerWord[];
  nonsensical_words: string[];
  pacing_timeline: PacingSegment[];
  ai_feedback: string;
}

export type GradingSourceType = 'self' | 'instructor';
export type GradingContextType = 'practice' | 'class';

export interface Grading {
  id: string;
  transcriptId: string;
  audioFileId: string | null;
  audioOwnerId: string | null;
  presentationTitle: string | null;
  rubricId: string | null;
  rubricName: string | null;
  status: 'processing' | 'completed' | 'failed';
  overallScore: number | null;
  maxPossibleScore: number | null;

  // Grading context fields
  sourceType: GradingSourceType;
  contextType: GradingContextType;
  contextId: string | null;  // class_id when contextType = "class"
  contextName: string | null;  // class name for display
  isOfficial: boolean;

  pacingWpmAvg: number | null;
  pacingWpmVariance: number | null;
  pacingPauseCount: number | null;
  pacingScore: number | null;

  clarityFillerWordCount: number | null;
  clarityFillerWordPercentage: number | null;
  clarityNonsensicalWordCount: number | null;
  clarityScore: number | null;

  detailedResults: GradingDetailedResults | null;
  gradedByUserId: string | null;
  gradedByName: string | null;
  gradedByRole: string | null;
  createdAt: string;
}

export interface RubricCriterionRequest {
  name: string;
  description: string;
  maxScore: number;
  weight: number;
}

export interface RubricCreateRequest {
  name: string;
  description?: string;
  criteria: RubricCriterionRequest[];
}

export interface RubricUpdateRequest {
  name?: string;
  description?: string;
  criteria?: RubricCriterionRequest[];
}

export interface GradingInitiateRequest {
  transcript_id: string;
  rubric_id: string;
  // Optional grading context fields
  source_type?: GradingSourceType;
  context_type?: GradingContextType;
  context_id?: string;
  is_official?: boolean;
}

File: backend/app/schemas/grading.py
from typing import Optional, List

from pydantic import BaseModel, Field


class GradingInitiateRequest(BaseModel):
    transcript_id: str
    rubric_id: str
    source_type: Optional[str] = None
    context_type: Optional[str] = None
    context_id: Optional[str] = None
    is_official: Optional[bool] = None


class GradingResponse(BaseModel):
    id: str
    transcriptId: str = Field(alias="transcriptId")
    audioFileId: Optional[str] = Field(alias="audioFileId")
    audioOwnerId: Optional[str] = Field(default=None, alias="audioOwnerId")
    presentationTitle: Optional[str] = Field(alias="presentationTitle")
    rubricId: Optional[str] = Field(alias="rubricId")
    rubricName: Optional[str] = Field(alias="rubricName")
    status: str
    overallScore: Optional[float] = Field(alias="overallScore")
    maxPossibleScore: Optional[float] = Field(alias="maxPossibleScore")
    sourceType: str = Field(default="self", alias="sourceType")
    contextType: str = Field(default="practice", alias="contextType")
    contextId: Optional[str] = Field(default=None, alias="contextId")
    contextName: Optional[str] = Field(default=None, alias="contextName")
    isOfficial: bool = Field(default=False, alias="isOfficial")
    pacingWpmAvg: Optional[float] = Field(alias="pacingWpmAvg")
    pacingWpmVariance: Optional[float] = Field(alias="pacingWpmVariance")
    pacingPauseCount: Optional[int] = Field(alias="pacingPauseCount")
    pacingScore: Optional[float] = Field(alias="pacingScore")
    clarityFillerWordCount: Optional[int] = Field(alias="clarityFillerWordCount")
    clarityFillerWordPercentage: Optional[float] = Field(alias="clarityFillerWordPercentage")
    clarityNonsensicalWordCount: Optional[int] = Field(alias="clarityNonsensicalWordCount")
    clarityScore: Optional[float] = Field(alias="clarityScore")
    detailedResults: Optional[dict] = Field(alias="detailedResults")
    gradedByUserId: Optional[str] = Field(default=None, alias="gradedByUserId")
    gradedByName: Optional[str] = Field(default=None, alias="gradedByName")
    gradedByRole: Optional[str] = Field(default=None, alias="gradedByRole")
    createdAt: str = Field(alias="createdAt")

    class Config:
        from_attributes = True
        populate_by_name = True
